src/sql/src/plan/typeconv.rs

```diff
@@ -360,13 +366,27 @@ static VALID_CASTS: Lazy<BTreeMap<(ScalarBaseType, ScalarBaseType), CastImpl>> =
 
         // TIMESTAMP
         (Timestamp, Date) => Assignment: CastTimestampToDate(func::CastTimestampToDate),
-        (Timestamp, TimestampTz) => Implicit: CastTimestampToTimestampTz(func::CastTimestampToTimestampTz),
+        (Timestamp, TimestampTz) => Implicit: CastTemplate::new(|_ecx, _ccx, _from_type, to_type| {
+            let p = to_type.unwrap_timestamp_precision();
+            Some(move |e: HirScalarExpr| e.call_unary(CastTimestampToTimestampTz(func::CastTimestampToTimestampTz(p))))
+        }),
+        (Timestamp, Timestamp) => Implicit: CastTemplate::new(|_ecx, _ccx, _from_type, to_type| {
```

Comment:
Oh, yes, great point. This should be an `Assignment` cast, not an implicit cast, to match how this works for `(Numeric, Numeric)`. I got this right for `(TimestampTz, TimestampTz)` but messed it up here.